kuka_rsi_hw_interface/src/rsi_command_handler.cpp

```diff
@@ -0,0 +1,325 @@
+
+#include <string>
+#include <cstring>
+#include <stdexcept>
+#include <iostream>
+
+#include "rsi_command_handler.hpp"
+
+namespace kuka_rsi_hw_interface
+{
+RSICommandHandler::RSICommandHandler()
+: state_data_structure_("Rob"), command_data_structure_("Sen"), err_buf_("")
+{
+  // Later this should be defined by the rsi xml
+  state_data_structure_.GetParams()["TYPE"] = xml::XMLParam(xml::XMLType::STRING);
+  // how to get string: std::get<XML::xml_string>(command_data_structure_.params_["TYPE"]).first
+  xml::XMLElement Out("Out");
+  Out.GetParams()["01"] = xml::XMLParam(xml::XMLType::BOOL);
+  Out.GetParams()["02"] = xml::XMLParam(xml::XMLType::BOOL);
+  Out.GetParams()["03"] = xml::XMLParam(xml::XMLType::BOOL);
+  Out.GetParams()["04"] = xml::XMLParam(xml::XMLType::BOOL);
+  Out.GetParams()["05"] = xml::XMLParam(xml::XMLType::BOOL);
+  xml::XMLElement Override("Override");
+  Override.GetParams()["Override"] = xml::XMLParam(xml::XMLType::LONG);
+  state_data_structure_.GetChilds().emplace_back(Out);
+  state_data_structure_.GetChilds().emplace_back(Override);
+
+  command_data_structure_.GetParams()["Type"] = xml::XMLParam(xml::XMLType::STRING);
+  xml::XMLElement RKorr("RKorr");
+  RKorr.GetParams()["X"] = xml::XMLParam(xml::XMLType::DOUBLE);
+  RKorr.GetParams()["Y"] = xml::XMLParam(xml::XMLType::DOUBLE);
+  RKorr.GetParams()["Z"] = xml::XMLParam(xml::XMLType::DOUBLE);
+  RKorr.GetParams()["A"] = xml::XMLParam(xml::XMLType::DOUBLE);
+  RKorr.GetParams()["B"] = xml::XMLParam(xml::XMLType::DOUBLE);
+  RKorr.GetParams()["C"] = xml::XMLParam(xml::XMLType::DOUBLE);
+  xml::XMLElement DiO("DiO");
+  DiO.GetParams()["DiO"] = xml::XMLParam(xml::XMLType::LONG);
+  command_data_structure_.GetChilds().emplace_back(RKorr);
+  command_data_structure_.GetChilds().emplace_back(DiO);
+}
+
+bool RSICommandHandler::Decode(char * const buffer, const size_t buffer_size)
+{
+  std::cout << "Decode started" << std::endl;
+  auto buffer_it = buffer;
+  try {
+    decodeNode(state_data_structure_, buffer, buffer_it, buffer_size);
+  } catch (const std::exception & e) {
+    std::cout << "### ERROR ###" << std::endl;
+    std::cerr << e.what() << '\n';
+    return false;
+  }
+  std::cout << "Decode finished" << std::endl;
+  return true;
+}
+
+bool RSICommandHandler::Encode(char * & buffer, const size_t buffer_size)
+{
+  std::cout << "Decode started" << std::endl;
+  auto buffer_it = buffer;
+  try {
+    encodeNode(command_data_structure_, buffer, buffer_it, buffer_size);
+  } catch (const std::exception & e) {
+    std::cout << "### ERROR ###" << std::endl;
+    std::cerr << e.what() << '\n';
+    return false;
+  }
+  std::cout << "Decode finished" << std::endl;
+  return true;
+}
+
+void RSICommandHandler::decodeNode(
+  xml::XMLElement & element, char * const buffer, char * & buffer_it,
+  const size_t buffer_size)
+{
+  xml::XMLString node_name(buffer, 0);
+  // fast forward for the first open bracket
+  for (; *buffer_it != '<'; buffer_it++) {
```

Comment:
might be done later, but a lot of trash can come between the end tag and the next < tag here
e.g. <abc a="1"/>hahaha&@{đ is valid according to the code
I think you were aware of this, but then this should be documented somewhere